aicssegmentation/workflow/workflow.py

```diff
@@ -162,3 +166,104 @@ def is_done(self) -> bool:
+    """
+
+    def __init__(
+        self, workflow_definition: WorkflowDefinition, input_dir: str, output_dir: str, channel_index: int = 0
+    ):
+        if workflow_definition is None:
+            raise ValueError("workflow_definition")
+        self._workflow_definition = workflow_definition
+
+        self.input_path = Path(input_dir)
+        if not self.input_path.exists():
+            raise ValueError("The input directory does not exist")
+
+        self.output_path = Path(output_dir)
+        # Creating an the output directory at output_dir if it does not exist already
+        if not self.output_path.exists():
+            if self.output_path.parent.exists():
+                self.output_path.mkdir(parents=True, exist_ok=True)
+            else:
+                raise ValueError(f"Output directory does not exist, and cannot be created at {self.output_path.parent}")
+
+        self.files_count: int = 0
+        self.failed_files: Dict[Path, str] = dict()
+        self._channel_index = channel_index
+
+    def is_valid_image(self, image_path: Path) -> bool:
+        """
+        Check if file at a given image_path is a valid image type we support (
+
+        Params:
+            image_path (Path): image to check
+
+        Returns:
+            (bool): True if all images are .tiff
+        """
+        if (
+            image_path.suffix.lower() == ".tiff"
+            or image_path.suffix.lower() == ".tif"
+            or image_path.suffix.lower() == ".czi"
+        ):
+            return True
+        else:
+            return False
+
+    def process_all(self):
+        """
+        Process all images in the input_dir with the workflow_definition used to set up the BatchWorkflow
+
+        Params:
+            none
+
+        Returns:
+            none
+        """
+        # Currently will save files in same format as they are in the input path
+        for f in listdir(self.input_path):
+            full_path = Path(self.input_path).joinpath(f)
+            self.files_count += 1
+            if self.is_valid_image(full_path):
+                # read and format image in the way we expect
+                image_from_path = imread(full_path).squeeze()
+                if image_from_path.ndim > 4:
+                    raise ValueError("Image is over 4 dims")
+                if image_from_path.ndim == 4:
+                    image_from_path = image_from_path[self._channel_index, :, :, :]
+                try:
+                    # Run workflow on image
+                    workflow = Workflow(self._workflow_definition, image_from_path)
+                    result = workflow.execute_all()
+                except Exception as e:
+                    # Handle failures during workflow execution/save
+                    self.failed_files[full_path] = f"Failed during processing with error {e}"
+            else:
+                self.failed_files[full_path] = f"Unsupported image type {full_path.suffix}"
+        self.write_log_file()
```

Comment:
Consider appending to the log file as files are processed instead of one big dump at the end. That way if something unrecoverable happens for whatever reason in the middle of a big batch - at least we have a partial log to look at.